service/dap/server.go

```diff
@@ -882,22 +882,30 @@ func (s *Server) onStepOutRequest(request *dap.StepOutRequest) {
 
 func (s *Server) doStepCommand(command string, threadId int) {
 	// Use SwitchGoroutine to change the current goroutine.
-	state, err := s.debugger.Command(&api.DebuggerCommand{Name: api.SwitchGoroutine, GoroutineID: threadId}, nil)
-	if err != nil {
-		s.log.Errorf("Error switching goroutines while stepping: %e", err)
-		// If we encounter an error, we will have to send a stopped event
-		// since we already sent the step response.
-		stopped := &dap.StoppedEvent{Event: *newEvent("stopped")}
-		stopped.Body.AllThreadsStopped = true
-		if state.SelectedGoroutine != nil {
-			stopped.Body.ThreadId = state.SelectedGoroutine.ID
-		} else if state.CurrentThread != nil {
-			stopped.Body.ThreadId = state.CurrentThread.GoroutineID
+	var err error
+	if threadId != 0 { // TODO: threadId == 0 could be valid the functions calling this should actually check whether or not the client specified a threadId argument.
+		_, err = s.debugger.Command(&api.DebuggerCommand{Name: api.SwitchGoroutine, GoroutineID: threadId}, nil)
+		if err != nil {
+			s.log.Errorf("Error switching goroutines while stepping: %e", err)
+			// If we encounter an error, we will have to send a stopped event
+			// since we already sent the step response.
+			state, err := s.debugger.State(false)
+			if err != nil {
+				s.log.Errorf("Error retrieving state: %e", err)
+				return
```

Comment:
You should not just return here on error. Then the client will think that you succeeded and are still running the step command - see comment on line 890. It will mark the thread as running in the stack trace. You must translate the error into a stopped event to let the client know that nothing is running.

Reply:
Done.